### ISSUE TITLE
full_triangle_ for BF method

### ISSUE THREAD

**Issue by mlbugs** [casact/chainladder-python#143]
BF method doesn't seem to be creating the runoff triangle correctly.  In my data, accident months can have volatile reported losses as of the latest valuation date.  Using an expected IBNR method can therefore produce volatile LDF's (implied, using ultimate_ / latest.diagonal ).  In my particular case, when reported losses are an outlier on the low end, the implied LDF is very high.  When I extract the first diagonal from the runoff triangle and divide that into my ultimate*, the resulting LDF's are much more homogeneous than they were the previous months.  Assuming my development pattern is relatively benign, there shouldn't be this much smoothing (development) in a one month projection.  For example, if the percent of ultimate for month _n_ is estimated to be 90% and for _n+1_ it's 95%, the LDF shouldn't drop from 3.0 to 1.2, it should drop to 1.6 (only 3.3% of ultimate has been reported for the month in questions as of month _n_).  The projected ultimate is only 13.3% and an expectation of 5% being reported next month would yield an expected reported loss of 8.3% and an implied ldf of 1.6.

Use case here is daily interpolation between month start and month end (ie, subsequent) LDF's.  It's better to interpolate to the expected month-end ldf than the subsequent accident month's ldf as this could produce large swings in your projected ultimate losses during the month.

*First diagonal from runoff triangle:
bf_table2 = bf_table.full_triangle_.dev_to_val()
bf_table3 = bf_table2[bf_table2.valuation==(triangle.valuation_date + relativedelta(months=+1))]
next_mo_ldf = bf_table.ultimate_ / bf_table3

for relativedelta, use "from dateutil.relativedelta import *"

**Comment by jbogaardt** [comment]
Hi @mlbugs , Let me state how it _currently_ works - The run-off for future valuations of a triangle is entirely a function of the estimators `ultimate_` vector and its `cdf_`.  Specifically it is `estimator.ultimate_ / estimator.cdf_`.  This "run-off" formula is shared across all estimators including the expected loss methods. See [code implementation](https://github.com/casact/chainladder-python/blob/master/chainladder/core/common.py#L19-L20).   

Are you suggesting that for expected loss methods, the run-off should also be a function of the previous diagonal as well?  

It would help me if we could work from a reproducible example.  Is it possible to reproduce the issue from this sample as a baseline as well as your expectation for the correct answer?

```python
import chainladder as cl

triangle = cl.load_sample('raa')
exposure = triangle.latest_diagonal*0 + 20000
bf_model = cl.BornhuetterFerguson(apriori=0.65)
bf_model.fit(triangle, sample_weight=exposure)

pct_ult = bf_model.full_triangle_ / bf_model.ultimate_
```

**Comment by mlbugs** [comment]
Hi John, 

Add in numpy and replace the 20000 with 

`np.random.normal(20000, 10000, size=(10,1))`

At this level of variance, the issue should manifest itself in at least one or two AY's.  

Compare the implied ldf's and the 1 year projection of the ldf's from:

`bf_model.ultimate_ / triangle.latest_diagonal`
with
```
bf_model2 = bf_model.full_triangle_.dev_to_val()
bf_model3 = bf_model2[bf_model2.valuation==(triangle.valuation_date + relativedelta(years=+1))]
bf_model.ultimate_ / bf_model3
```
There should be at least one year where the 'forecasted' ldf swings dramatically from where the current one is.  Or you can just examine the `pct_ult` array and there should be one or two large discontinuities between the latest diagonal and the first expected diagonal.  If it's not apparent, rerun the simulation and it should eventually turn up.  

For expected loss methods, I'd postulate that the runoff triangle should be a function of the expected losses (and the estimated percent of ultimate), instead of the ultimate losses and pct_ult.  That's more in sync with how expected loss methods work but I understand that probably complicates the full_triangle_ function.

Morgan

**Comment by jbogaardt** [comment]
I think I see what you're saying. In this example, AY1988 looks like it has weird run-off.
```python
import chainladder as cl
import numpy as np
np.random.seed(100)

triangle = cl.load_sample('raa')
exposure = triangle.latest_diagonal*0 + np.random.normal(20000, 10000, size=(10,1))
bf_model = cl.BornhuetterFerguson(apriori=0.65)
bf_model.fit(triangle, sample_weight=exposure)

pct_ult = bf_model.full_triangle_ / bf_model.ultimate_
pct_ult[pct_ult.valuation>=triangle.valuation_date].iloc[..., :-1].T.plot(
    title='Latest diagonal and later');
```

I need to think on this one a bit more. I'd like to find a general formula that works for both the expected loss methods and the development factor methods.  This is important because I want to blend different kinds of methods using a `VotingChainladder` and would like the run-off from the blended methods to also be producible.

**Comment by mlbugs** [comment]
The next easiest thing I can think of would be to do something like taking (ultimate_ - latest_diagonal) for both the losses and pct_ult.  Normalize the remaining pct_ult so it ranges from (0-1] for the runoff months and then use that to spread the IBNR.  Not the greatest approach but should still work across all methods.  Haven't familiarized myself with VCL yet, so I'm not sure what wrinkles that presents.  Will take a look.

**Comment by jbogaardt** [comment]
How is this solution? Think these are the right numbers? If so, I can probably get a generalized form in.
```python
import chainladder as cl
import numpy as np
import matplotlib.pyplot as plt

fig, ((ax0, ax1)) = plt.subplots(ncols=2, figsize=(15,5))
np.random.seed(100)

triangle = cl.load_sample('raa')
exposure = triangle.latest_diagonal*0 + np.random.normal(20000, 10000, size=(10,1))
bf_model = cl.BornhuetterFerguson(apriori=0.65)
bf_model.fit(triangle, sample_weight=exposure)

# original
pct_ult = bf_model.full_triangle_ / bf_model.ultimate_
pct_ult[pct_ult.valuation>=triangle.valuation_date].iloc[..., :-1].T.plot(
    title='Latest diagonal and later (Current)', ax=ax0);

# Revised approach
new_run_off = (bf_model.full_expectation_ / bf_model.ultimate_ *.65*exposure)
new_run_off = new_run_off[new_run_off.valuation>=triangle.valuation_date].cum_to_incr()
new_run_off = new_run_off[new_run_off.valuation > triangle.valuation_date].incr_to_cum() + triangle.latest_diagonal
new_run_off = new_run_off[new_run_off.valuation > triangle.valuation_date] + triangle
new_pct_ult = (new_run_off / bf_model.ultimate_)

new_pct_ult[new_pct_ult.valuation>=triangle.valuation_date].iloc[..., :-1].T.plot(
    title='Latest diagonal and later (Revised)', ax=ax1);
```
![image](https://user-images.githubusercontent.com/19401133/113954638-6f89d980-97d7-11eb-8724-687a69dea8de.png)

**Comment by mlbugs** [comment]
This looks like it works - thanks, John!  Maybe make this as an option for the function?

**Comment by jbogaardt** [comment]
Cool. For the development factor approaches, this seems to produce identical results to the current implementation when substituting the ultimate vector for the expected (`bf_model.apriori * exposure` with `cl_model.ultimate_`). That makes me confident that it is the right generalized implementation.

**Comment by jbogaardt** [comment]
Added this update to the main branch. Will show up in the 0.8.3 release.  If you want to play with it ahead of that, feel free to install from git.

**Comment by mlbugs** [comment]
Thanks, John!

Seems like this could be an area of reserving research, particularly if the
reserve forecast metrics from Richman's recent paper are added to
chainladder and researchers start to scrutinize backested AvE for the
accuracy of reserving methods.

**Comment by mlbugs** [comment]
Thanks for releasing the 8.3!  

While poking around the CAS exam syllabi for something unrelated, I came across a reference for expected loss emergence methods (pg 7), with method 2 being the original implementation you had:  https://www.casact.org/sites/default/files/2021-03/7_Clark.pdf.

I suppose a critique of my proposed approach is that it could potentially produce biased estimates of the expected loss emergence because it makes the expected loss emergence a function of the actual loss.  But like you said, at least in the case of chainladder, the ultimate is already a function of the actual, thereby making the expected emergence a function of the actual (ie, your comment re: the new formula generalizing the chain ladder approach).

**Comment by jbogaardt** [comment]
And thank you for the suggested improvement @mlbugs.  To be clear on your last comment, do you think there is more work to be done on this one?  I convinced myself the new solution is the best generalized approach, but you've cast some doubt on me.
Here is what I like about the `0.8.3` solution. Consider the comparison of the Expected Loss method, the Bornhuetter Ferguson method and the Chainladder method.

![image](https://user-images.githubusercontent.com/19401133/116020230-1c02f280-a603-11eb-9774-96d151fff49d.png)

1. Expected Loss method is entirely based on the expected loss and the known data has no influence on the lower half of the `full_triangle_`.  This is the "Method 1" in the Clark paper, BUT it only gets used on this one method. Because it cares not of the actual data, there is a huge step change from the latest diagonal to the first unreported diagonal.
2. BornhuetterFerguson is a credibility weighting of latest data and an apriori belief.  This was your suggestion and it fits the generalized approach well. Not only that, it extends to any `Benktander.n_iters` options. As `n_iters` increases, the run-off moves closer to the "Method 2" of the Clark paper. Likewise, as it decreases it approaches "Method 1"
3. Finally, the Chainladder follows the "Method 2" of the Clark paper, BUT, like (1),  it only gets used on this one method.

In all cases, I've not touched the `full_expectation_` property. It is akin to "Method 2" in the Clark paper and will produce a different lower half of the triangle than the `full_triangle_` property for anything other than the Expected Loss or Chainladder methods.  I'm unsure whether to revise this one.

I've struggled to find much of any literature on run-off for the different methods, so thanks for pointing out the Clark paper.  However, it feels like that paper misses the generalizations I describe in (1 - 3). In `0.8.3` it isn't either "Method 1" or "Method 2", but shades of those two approaches. While I like the mathematical properties of the implemented approach, I also want to avoid falling in love with it if it suffers from biases. It's received no significant peer review - so if you have time, give it some more thought. I'd be happy to patch it if there are unintended consequences in the approach.

Code for the plots:
```python
import chainladder as cl
import numpy as np
import matplotlib.pyplot as plt

fig, ((ax0, ax1, ax2)) = plt.subplots(ncols=3, figsize=(20,5))
np.random.seed(100)

triangle = cl.load_sample('raa')
exposure = triangle.latest_diagonal*0 + np.random.normal(20000, 10000, size=(10,1))
bk0_model = cl.Benktander(apriori=0.65, n_iters=0)
bk1_model = cl.Benktander(apriori=0.65, n_iters=1)
bk100_model = cl.Benktander(apriori=0.65, n_iters=100)
bk0_model.fit(triangle, sample_weight=exposure)
bk1_model.fit(triangle, sample_weight=exposure)
bk100_model.fit(triangle, sample_weight=exposure)

pct_ult = bk0_model.full_triangle_ / bk0_model.ultimate_
pct_ult[pct_ult.valuation>=triangle.valuation_date].iloc[..., :-1].T.plot(
    title='Expected Loss Method', ax=ax0);
pct_ult = bk1_model.full_triangle_ / bk1_model.ultimate_
pct_ult[pct_ult.valuation>=triangle.valuation_date].iloc[..., :-1].T.plot(
    title='Bornhuetter Ferguson', ax=ax1);
pct_ult = bk100_model.full_triangle_ / bk100_model.ultimate_
pct_ult[pct_ult.valuation>=triangle.valuation_date].iloc[..., :-1].T.plot(
    title='Chainladder', ax=ax2);
```

**Comment by mlbugs** [comment]
Hey John,
Sorry, it took me a contrived example in excel to realize what you were saying:  my formula gives a generalization of expected loss emergence for both chainladder and BF methods.  Initially, I wasn't thinking my formula would exactly recreate the emergence pattern for expected loss methods/BF/CC.  But now I see it:  dividing the IBNR by (1-LDFi^-1) where LDFi^-1 is the percent emerged for the latest month of actuals for the given origin period yields the expected loss.  And in your prior message (from almost 3 weeks ago), you had pointed out that dividing by this expression for chainladder methods yields the CL ultimate loss indication.  

I ran your code and added a bk with n_iter=2 to help wrap my head around it.  With any number of iterations in BK, you're always working with the same loss emergence pattern.  The only difference is your 'a priori' loss ratio for a given iteration.  So once you've got a method that adapts the emergence pattern to your expected IBNR, you're set, and in retrospect, that's how I mashed it up in my head.

One thing your example did make me check was to add an expected loss method to my spreadsheet of contrived examples.  It looks like my method doesn't exactly recreate the emergence pattern generated as would otherwise be implied by this method.  For the expected loss method, you really don't care about the reported to date but I can't see a way for my method to escape it (will send  you a copy of my see separately).  That being said, I still like my method better as it is more robust to volatility in the actual reported losses.

I don't think any further work on your part is necessary and might have to try my hand at writing it up for e-Forum.

Thanks for all your help on this!!
